### DISCUSSION TITLE
Handling Time out exceptions

### DISCUSSION THREAD

**Discussion by shaanthomas** [scrapli/nornir_scrapli#140] [General]
I am experiencing time out issues with nornir-scrapli while using jinja templates. I have a script that loads in variables from a yaml file, renders a template using jinja2, and send the config to a device using nornir-scrapli. The first two tasks are working well, however the last task is not. I am receiving the Scrapli Timeoute error. After much reasearch, I think I need to do the following: leave the timeout_socket alone, set timeout_transport to 0, then just set timeout_ops to however long it takes for your longest output to print out on the terminal. The problem is that I do not know how to set these options in nornir. Is it done in the extras field on the hosts.yaml file?
I have attached screenshots of the hosts.yaml file, script, and the terminal. Also, I have attached the Nornir log file.

[nornir.log](https://github.com/scrapli/nornir_scrapli/files/11150554/nornir.log)
![script](https://user-images.githubusercontent.com/65083771/229850567-df8e17f2-0e58-4600-9f70-ab1c57ab41a9.png)
![hosts](https://user-images.githubusercontent.com/65083771/229850589-8c6df107-b5be-49fa-b63a-1b470370d653.png)
![exceptions](https://user-images.githubusercontent.com/65083771/229850601-7e19c521-f212-4c58-8aec-25e8d2febfe0.png)

Thank you in advance for your help.
Rashaan

**Comment by carlmontanari** [comment]
> I think I need to do the following: leave the timeout_socket alone, set timeout_transport to 0, then just set timeout_ops to however long it takes for your longest output to print out on the terminal. The problem is that I do not know how to set these options in nornir. Is it done in the extras field on the hosts.yaml file?

Seems reasonable, and yes, in extras. Like here: https://github.com/scrapli/nornir_scrapli/blob/main/examples/basic_netconf_usage/nornir_data/groups.yaml#L7-L9

you can pass whatever valid scrapli args in extras.

**Reply by shaanthomas** [reply]
Carl,
Thank you for your help. I will add the changes to the group file once it works with the group file. Do you recommend changing the hosts file to something that looks like this.

![hosts](https://user-images.githubusercontent.com/65083771/229856005-d11aa669-61bf-479c-b80a-19dd2a021f53.png)

**Comment by shaanthomas** [comment]
Carl,
Sorry to bug you again. I am still having issues. I added the timeout_transport:0 and timeout_ops: 60 variables under the data/extra key in the host.yaml file. I kept getting a noneType error.:
 File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/nornir/plugins/inventory/simple.py", line 26, in _get_connection_options
    hostname=c.get("hostname"),
Next. I looked at the link you sent and noticed that all the info was on the group.yaml file. I ammended my groups.yaml file. I am still getting the time out error.I have tried using multiple steps of 10. I am only changing the time_out ops variable. I have scoured the documentation. Am I missing a plugin? I thought this had something to do with the fact that I am using jinja templates with logic in the template (for loops). I took the for loop out of the template and received the same Send config error. Do you have anymore suggestions. Thanks for your quick response. I have included the hosts, group, te
[nornir.log](https://github.com/scrapli/nornir_scrapli/files/11151436/nornir.log)

![template](https://user-images.githubusercontent.com/65083771/229875840-53324948-d4b8-4a2a-9938-6fc0ff7d9d24.png)

![script](https://user-images.githubusercontent.com/65083771/229875830-1eb5e2e5-77ee-4658-9cda-7bd7d6be2560.png)
mplate, and log file.
I really appreciate your help.
![terminal output](https://user-images.githubusercontent.com/65083771/229875815-1097d4f0-2b1d-4d94-a64d-11e40ff741b2.png)

![groups](https://user-images.githubusercontent.com/65083771/229875803-1793e35a-ce4b-4dd6-9a1f-5c6df950208d.png)

**Reply by carlmontanari** [reply]
grab scrapli logs to see whats going on https://github.com/carlmontanari/scrapli/tree/main/examples/logging

**Comment by shaanthomas** [comment]
Scrapli and Nornir attached. It seems to be connecting to the switch.. which is an CSR1, but it is only reading \n which is next lines.....?


[scrapli.log](https://github.com/scrapli/nornir_scrapli/files/11151646/scrapli.log)
![Scrapli log screenshot](https://user-images.githubusercontent.com/65083771/229881359-7c4bea1f-fd90-4ec1-9b4b-b939486565e8.png)

[nornir.log](https://github.com/scrapli/nornir_scrapli/files/11151657/nornir.log)

**Reply by carlmontanari** [reply]
looks like whatever you are sending in configs is not what you think it is.

**Reply by shaanthomas** [reply]
Carl,
Sorry, I had to put out another fire. But I checked out the configs. Configs is a list..which is what i think should be. But there is something strange at the beginning of the list. I think it is some kind of encoding done by the jinja2 template. The '\ufeff\ufeff\' thing. Have you ever come across this?

['\ufeff\ufeff', 'interface loopback100', ' ip address 100.1.1.1 255.255.255.0', '', 'interface loopback11', ' ip address 11.1.1.0 255.255.255.0', '', 'interface GigabitEthernet2', ' ip address 1.1.1.1 255.255.255.0', ' ip ospf network point-to-point', ' no shutdown', ' negotiation auto', ' no mop enabled', ' no mop sysid', '', 'ip forward-protocol nd', 'ip tcp synwait-time 10', 'no ip http server', 'ip http authentication local', 'ip http secure-server', '', 'service tcp-keepalives-in', 'service tcp-keepalives-out', 'service timestamps debug datetime msec', 'service timestamps log datetime msec', 'service password-encryption', 'service compress-config', 'platform qfp utilization monitor load 80', 'platform punt-keepalive disable-kernel-core', 'no ip gratuitous-arps', 'no ip domain lookup', '', 'end']

**Comment by shaanthomas** [comment]
Carl,
I was able to fix it. The device was received the \ufeff\ufeff', before the config. I think its called a BOM. The device was not able to read it.  This is encoding that is append by the Jinja 2 template depending on where you create the template. My template was created in vS Code and you can change the encode settings on VS Code . I ended up just replacing the 'ufeff\ with nothing. I used the replace method. Thanks for your help
